### ISSUE TITLE
CSV count wrong when new line in cell

### ISSUE THREAD

**Issue by D-system** [Shopify/maintenance_tasks#977]
When a CSV has new lines inside a cell, the default count is incorrect.

CSV sample:
```csv
title,tags
"foo title","elephant
zebra
giraffe"
"bar title","one"

```

Representation:
<img width="325" alt="SCR-20240220-mzuc" src="https://github.com/Shopify/maintenance_tasks/assets/853015/d4164b1e-7548-4d34-bbff-7796793edb93">

There are 2 content rows but with the current implementation, it is shown as 4 rows.
The current implementation use `"\n"` as divider:
https://github.com/Shopify/maintenance_tasks/blob/a9ed8f13fbf9069e93a44882f0c9af2d8c7bec24/app/models/maintenance_tasks/batch_csv_collection_builder.rb#L29-L39

The precessing of the rows themselves seems to be correct (precessing 2 rows).

The task is shown as finished with `Processed 2 out of 4 items (50%).`
<img width="615" alt="SCR-20240220-ncpv" src="https://github.com/Shopify/maintenance_tasks/assets/853015/01eeff0b-b4c0-41c0-95a5-2ca4538387ad">

**Comment by D-system** [comment]
@etiennebarrie Merci Étienne pour l'explication.
I'm puzzled because at some point, the CSV needs to be parsed before to be given to the `process` method, isn't it? 
Could Maintenance Tasks do or update the count at that time?

**Comment by etiennebarrie** [comment]
Yes, the CSV is parsed by job-iteration at some point, but this is done as a stream, just reading enough data to emit a row. So we don't know how many rows there are until we're finished basically.

We could change the display a bit, but we do handle all the different cases:
<img width="1093" alt="image" src="https://github.com/Shopify/maintenance_tasks/assets/3535/b8fa3827-fef8-45a6-9e8a-52275fda83f2">
These are multiple runs of UpdatePostsTask, when `count` is defined as returning respectively 8, `nil`, 12, and 10. There are always 10 Post records.

So when `count` is not defined in the task, and it's the default implementation for CSV, we can have an incorrect count, but then we handle it like we would do if you had provided an incorrect count.

One solution would be to set the total to the number of rows we've found at the end of a run, but then we would lose some information, namely the expected count that was defined by the task (implicitly or explicitly).

We could also provide an option to `csv_collection`, e.g. `csv_collection exact_count: true` but that's not great.

At the very least, we should add this in the README, in the section about CSV. We mention the trailing newline, but don't explain why (it's in order to have the correct count, when rows don't include newlines). We could expand and explain that the count is based on the number of lines and will be off if rows includes newlines or if the trailing newline is missing, and maybe even give a one-liner to get correct counts for those cases.

```ruby
  def count = CSV.parse(csv_content, headers: true).count
```

---

Taking a step back, given that we've moved to automatically count the number of records for Active Record relations in v2.0.0, and that can also be very slow, it could be argued that it's fair to make the CSV implicit counting slower but correct by default, and leave it to task authors to override count with `nil` or a faster implementation (if they don't expect newlines in rows) if they really need to.

@adrianna-chang-shopify @nvasilevski  WDYT? Would that be too much of a breaking change for a minor bump? I.e. a 10-million-line CSV would cause the run to take 20 seconds at the beginning of the first job to count the lines. I think I'm in favor of just documenting this in the README.

**Comment by D-system** [comment]
Thank you for taking the time to explain. I didn't know it was streaming the CSV.
I'm happy with an update of the readme and/or update the task template with the accurate count one-liner method commented.

**Comment by etiennebarrie** [comment]
> I'm happy with an update of the readme

Please do!

> update the task template

But not that part. 😄 If we feel display correctness is more important, then we should fix it, but we shouldn't just cause more code to be in all apps.

---

In the readme, we should mention that the count is only an approximation counting the newlines in the CSV, and that if they have newlines in their CSV and want an accurate total, they can override `count`. My understanding is that most maintenance tasks CSVs are full of ids instead, and won't need that, so it shouldn't be in the template.

**Comment by D-system** [comment]
I made a PR to update the `#count` method: https://github.com/Shopify/maintenance_tasks/pull/991
I was not sure about updating for README too, so I made a different commit to revert it in the case it is not wanted.
